sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala

```diff
@@ -314,6 +314,20 @@ object NestedColumnAliasing {
  */
 object GeneratorNestedColumnAliasing {
   def unapply(plan: LogicalPlan): Option[LogicalPlan] = plan match {
+    // Either `nestedPruningOnExpressions` or `nestedSchemaPruningEnabled` is enabled, we
+    // need to prune nested columns through Project with under Filter then under Generate.
+    // The difference is when `nestedSchemaPruningEnabled` is on, nested columns will
+    // be pruned further at file format readers if it is supported.
+    case Project(projectList, Filter(condition, g: Generate))
+      if (SQLConf.get.nestedPruningOnExpressions ||
+        SQLConf.get.nestedSchemaPruningEnabled) && canPruneGenerator(g.generator) =>
```

Comment:
So this basically adds a new case pattern, the pruning logic is the same?

Reply:
> So this basically adds a new case pattern, the pruning logic is the same?

Same like `NestedColumnAliasing` support Project -> Filter -> child